Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -0,0 +1,381 @@
+        // Registers the SwitcherEventHandler which is defined in ComponentModeSwitcher
+        ViewportUi::ViewportUiRequestBus::Event(
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::RegisterSwitcherEventHandler,
+            m_switcherId,
+            m_handler);
+    }
+
+    ComponentModeSwitcher::~ComponentModeSwitcher()
+    {
+        ViewportEditorModeNotificationsBus::Handler::BusDisconnect();
+        EntityCompositionNotificationBus::Handler::BusDisconnect();
+        ToolsApplicationNotificationBus::Handler::BusDisconnect();
+
+        ViewportUi::ViewportUiRequestBus::Event(
+            ViewportUi::DefaultViewportId, &ViewportUi::ViewportUiRequestBus::Events::RemoveSwitcher, m_switcherId);
+    }
+
+    void ComponentModeSwitcher::UpdateSwitcherOnEntitySelectionChange(
+        [[maybe_unused]] const EntityIdList newlySelectedEntityIds, [[maybe_unused]] const EntityIdList newlyDeselectedEntityIds)
+    {
+        AZ::Entity* entity = nullptr;
+
+        // Currently only handling one new entity selection at a time
+        if (!newlySelectedEntityIds.empty() && newlySelectedEntityIds.size() == 1)
+        {
+            AZ::ComponentApplicationBus::BroadcastResult(
+                entity, &AZ::ComponentApplicationBus::Events::FindEntity, AZ::EntityId(newlySelectedEntityIds.front()));
+
+            if (entity)
+            {
+                EntityIdList selectedEntityIds;
+                ToolsApplicationRequestBus::BroadcastResult(selectedEntityIds, &ToolsApplicationRequestBus::Events::GetSelectedEntities);
+        
+                if (selectedEntityIds.size() > 1)
+                {
+                    RemoveExclusiveComponents(*entity);
+                    return;
+                }
+
+                for (const auto& entityComponent : entity->GetComponents())
+                {
+                    // Find if the component has a component mode using GetNumOfEventHandlers
+                    const auto entityComponentIdPair = AZ::EntityComponentIdPair(entity->GetId(), entityComponent->GetId());
+                    AddComponentButton(entityComponentIdPair);
+                }
+            }
+        }
+        else if (!newlyDeselectedEntityIds.empty())
+        {
+        /*    AZStd::vector<AZ::EntityComponentIdPair> pairIdsToRemove;
+            for (auto componentDataIt : m_addedComponents)
+            {
+                pairIdsToRemove.push_back(componentDataIt.m_pairId);
+            }
+            for (auto pairId : pairIdsToRemove)
+            {
+                RemoveComponentButton(pairId);
+            }*/
+            //v.erase(std::remove_if(v.begin(), v.end(), is_odd), v.end());
+            for (AZStd::vector<ComponentData>::reverse_iterator it = m_addedComponents.rbegin(); it != m_addedComponents.rend(); ++it)
+            {
+                RemoveComponentButton(it->m_pairId);
+            }
```

Comment:
suggestion: It might be nice to have a `ClearComponentButtons` member function that calls this logic to make the API a little clearer and we can then reuse it in other places potentially